### PR TITLE
Article: On Growing Great Developers

### PR DESCRIPTION
**Pull request by adarsh**
Addresses #53

### CODE REVIEW

**Comment by adarsh** [comment]
cc @captbaritone, @gabebw, @lilliealbert, @shamus for feedback!

File: source/articles/2016/on-growing-great-developers.html.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+* Encourage candidates to network, go to meet ups, etc. Community is part of the job.
+
+## Shameless Plug
+
+Do you want help setting up a program like this?
+[Email me](mailto:hi@cylinder.work?subject=Saw Your Blog Post! Can you Help Us With Our Apprenticeship Program?)!
+
+## Acknowledgements and Further Resources
+
+* Thanks to [Jeremy Morony][2], who shared back his notes when we discussed
```
Comment (captbaritone) [inline source/articles/2016/on-growing-great-developers.html.md]: "shared back"? Maybe just "who shared his notes with me after we discussed"?

File: source/articles/2016/on-growing-great-developers.html.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+* Set a decision point at the end of two months when you make full-time offers.
+  This gives the candidate time to look for another job.
+
+
+## Other Notes
+
+* Have retrospectives with everyone (Candidates, Apprentices, Mentors, Everyone
+  Else) as you go.
+* Frequently iterate based on the feedback.
+  Your success metric is "number of great full-time hires made out of the
```
Comment (captbaritone) [inline source/articles/2016/on-growing-great-developers.html.md]: Maybe:

> You success metric "should be"

File: source/articles/2016/on-growing-great-developers.html.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+  program."
+* Make sure candidates get the same experience they'd get on the job.
+  Helps you evaluate them and helps them evaluate you as an employer.
```
Comment (captbaritone) [inline source/articles/2016/on-growing-great-developers.html.md]: > "This" will help you evaluate

File: source/articles/2016/on-growing-great-developers.html.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+* Put up a separate landing page which feeds into your full-time hiring
+  pipeline. Run ads against it or blog/tweet about it.
+* Select for humility _and_ confidence. You want candidates are willing to take a
+  step back to learn, regardless of their past work history or accomplishments.
+* Only take candidates excited about the apprenticeship opportunity. Focus on
+  people who want to learn/grow. Otherwise you're selling something to a
+  reluctant buyer.
+* Respond quickly to applicants! You should do this for other candidates also.
+  Just sayin'.
```
Comment (captbaritone) [inline source/articles/2016/on-growing-great-developers.html.md]: 👍

File: source/articles/2016/on-growing-great-developers.html.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+GitHub Pages or Heroku." I gave this homework to lots of non-coding applicants.
+
+She was the only one to ever respond, and just four weeks later.
+Now I had two data points: `(0,0)` and
+`(4 weeks, learn enough front-end to make a portfolio site)`.
+
+Solving for a non-numerical `m` value,
+I guessed she would be ready for full-time work in 3 months.
+Turns out, it was 4 months, but her slope kept constant (or improved!). So as a
+result, I got a top-notch teammate and culture for a pretty small.
```
Comment (captbaritone) [inline source/articles/2016/on-growing-great-developers.html.md]: Pretty small what?

File: source/articles/2016/on-growing-great-developers.html.md [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+"If I coach this person for 3 months,
+would their answer meet our expectations for a full-time junior developer?"
+
+Now this involves a bit of imagination of what someone's learning vs. time slope
+is before you get to work with them.
+
+
+### What the Heck is a Learning vs. Time Slope?
+
+Through experience, I discovered each person has an internal rate of learning.
```
Comment (captbaritone) [inline source/articles/2016/on-growing-great-developers.html.md]: > I "have" discovered "that" each person

File: source/articles/2016/on-growing-great-developers.html.md [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+That is they can accumulate and apply new concepts at a certain rate.
+Imagine it looks like this:
+
+![skill vs. time](skill_vs_time.png)
+
+> _For any given candidate their slope will remain constant over time,
+given ongoing support. (AKA "Pandit's Postulate")_
+
+You use this imaginary non-quantifyable slope to both judge if they are good
```
Comment (captbaritone) [inline source/articles/2016/on-growing-great-developers.html.md]: Maybe:

> You "can" use

Or 

> Use this imaginary

**Comment by captbaritone** [comment]
Great post! Some questions I had:

* How should you decide how much to pay an apprentice?
* Do you have any experience with how a hiring structure like this works from a legal HR perspective?
* Any thoughts/experience on finding mentors or how to prevent an influx of jr engineers from distracting mentors from coding tasks?

**Comment by adarsh** [comment]
Cc @truffaut for review.

File: source/articles/2016/on-growing-great-developers.html.md [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+* You can buy your way out of the problem.
+* It is how everyone has hired for as long as anyone can remember.
+* Short-term impact. Expert developers can get ramped up quickly and help you ship much faster.
+
+### Why Farm
+
+* Cost - Creating a developer growth pipeline is much cheaper in the long-term. Apprentices are cheaper than experienced developers and will take a salary hit for the learning opportunity.
+* Loyalty - Apprentices are extremely loyal and less likely to turn over. You
+  took a chance on them and they will always remember.
+* Diversity - The experienced developer pool is very homogenous. The junior pool
```
Comment (gabebw) [inline source/articles/2016/on-growing-great-developers.html.md]: s/homogenous/homogeneous

File: source/articles/2016/on-growing-great-developers.html.md [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+  is not. If you can turn Juniors into Seniors, diversity will improve.
```
Comment (gabebw) [inline source/articles/2016/on-growing-great-developers.html.md]: "junior" in "junior pool" isn't capitalized, but "Junior" is capitalized here.

File: source/articles/2016/on-growing-great-developers.html.md [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+
+### Use the same interview process as your full-timers
+
+Give them the same questions/challenges as your full-timers
+but evaluate their responses with this lens:
+
+"If I coach this person for 3 months,
+would their answer meet our expectations for a full-time junior developer?"
+
+Now this involves a bit of imagination of what someone's learning vs. time slope
```
Comment (gabebw) [inline source/articles/2016/on-growing-great-developers.html.md]: I think I'd remove "Now" from the beginning of this sentence.

File: source/articles/2016/on-growing-great-developers.html.md [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+is before you get to work with them.
+
+
+### What the Heck is a Learning vs. Time Slope?
+
+Through experience, I discovered each person has an internal rate of learning.
+That is they can accumulate and apply new concepts at a certain rate.
```
Comment (gabebw) [inline source/articles/2016/on-growing-great-developers.html.md]: I'd add a comma after "is".

File: source/articles/2016/on-growing-great-developers.html.md (hunk 7 above)
Comment (gabebw) [inline source/articles/2016/on-growing-great-developers.html.md]: s/quantifyable/quantifiable

File: source/articles/2016/on-growing-great-developers.html.md [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+![skill vs. time](skill_vs_time.png)
+
+> _For any given candidate their slope will remain constant over time,
+given ongoing support. (AKA "Pandit's Postulate")_
+
+You use this imaginary non-quantifyable slope to both judge if they are good
+apprenticeship candidates, and later, if they would be good full-time hires.
+
+The trick here is to use an interview to figure out what the candidate's slope
+is. Good ole `y = mx + b`.
```
Comment (gabebw) [inline source/articles/2016/on-growing-great-developers.html.md]: s/ole/ol'

**Comment by adarsh** [comment]
@captbaritone thanks for feedback! Your questions:

> How should you decide how much to pay an apprentice?

Hrm. Good question. "Less than a full-timer but more than zero"? I don't know that I have a better answer. Do you think this is helpful to include in the article?

> Do you have any experience with how a hiring structure like this works from a legal HR perspective?

Honestly, not really. It might have been structured as a 1099 contracting arrangement, but I can't recall. I know thoughtbot decided to cover healthcare later on, since this was an important factor for those with families.

This may also impact the above question also, there may be minimum wage or other laws I'm not aware of.

> Any thoughts/experience on finding mentors or how to prevent an influx of jr engineers from distracting mentors from coding tasks?

Oh this is important. First off, it's important to not have too many apprentices. Maybe 1:10 or 1:20. Also, before you even start the program, the mentors should see the Jrs as an accelerator over time, as opposed to a burden or "distraction". You should at least have a few mentors that are very excited about the program, otherwise you're forcing something on the team which they may resent.

Added this into the article in 7358909.

**Comment by captbaritone** [comment]
> Less than a full-timer but more than zero

No, I suppose putting that in the article would not add much value 😀

**Comment by adarsh** [comment]
Thanks all for feedback! I'm going to ship it!

do it live

**Pull request merged**